### PR TITLE
Refactor

### PR DESCRIPTION
**Pull request by ljleppan**

### CODE REVIEW

File: src/main/java/fi/helsinki/cs/tmc/core/cache/ExerciseChecksumFileCache.java [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+
+    private Map<String, Map<String, String>> readCacheFile(Path file) throws TmcCoreException {
+        String oldCacheData = readCacheData(file);
+        Map<String, Map<String, String>> checksums = new HashMap<>();
+        if (oldCacheData != null && !oldCacheData.isEmpty()) {
+
+            try {
+                checksums = parser.fromJson(oldCacheData, TYPE_TOKEN);
+            } catch (JsonSyntaxException ex) {
+                System.err.println("WARNING: corrupt cache file, ignoring and overwriting");
```
Comment (jamo) [inline src/main/java/fi/helsinki/cs/tmc/core/cache/ExerciseChecksumFileCache.java:103]: logger?

Comment (ljleppan) [inline src/main/java/fi/helsinki/cs/tmc/core/cache/ExerciseChecksumFileCache.java:103]: Yes, just haven't gotten around to adding logging to TMC-Core yet. It's on the TODO :)

File: src/main/java/fi/helsinki/cs/tmc/core/communication/UrlCommunicator.java [hunk 2]
```diff
@@ -149,11 +149,10 @@ private MultipartEntityBuilder addParamsToRequest(
      * Tries to make GET-request to specific url.
      *
      * @param url URL to make request to
-     * @param params Any amount of parameters for the request. params[0] is always username:password
      * @return A Result-object with some data and a state of success or fail
```
Comment (jamo) [inline src/main/java/fi/helsinki/cs/tmc/core/communication/UrlCommunicator.java:154]: Maybe like:
Tries to make a GET-request and returns object representing the result.

File: src/main/java/fi/helsinki/cs/tmc/core/communication/UrlHelper.java [hunk 3]
```diff
@@ -3,12 +3,15 @@
...
+import java.net.URISyntaxException;
+
 public class UrlHelper {
 
     public final String apiParam;
     public final String clientParam;
     public final String coursesExtension;
-    public final String authExtension;
     private final TmcSettings settings;
     private String serverAddressPattern = "(https://)?([a-z]+\\.){2,}[a-z]+(/[a-z]+)*";
```
Comment (jamo) [inline src/main/java/fi/helsinki/cs/tmc/core/communication/UrlHelper.java]: we really should allow just http and make the variable private static final..

File: src/main/java/fi/helsinki/cs/tmc/core/communication/UrlHelper.java [hunk 4]
```diff
@@ -17,31 +20,27 @@ public UrlHelper(TmcSettings settings) {
...
-            url = url.substring(0, url.indexOf("?" + apiParam)) + params;
-        } else if (!url.endsWith(params)) {
-            url += params;
-        }
-        return url;
+    public String withParams(String url) throws URISyntaxException {
+        return new URIBuilder(url)
+                .addParameter(TmcConstants.API_VERSION_PARAM, settings.apiVersion())
+                .addParameter(TmcConstants.CLIENT_NAME_PARAM, settings.clientName())
+                .addParameter(TmcConstants.CLIENT_VERSION_PARAM, settings.clientVersion())
```
Comment (jamo) [inline src/main/java/fi/helsinki/cs/tmc/core/communication/UrlHelper.java]: maybe static import the constants

File: src/main/java/fi/helsinki/cs/tmc/core/communication/updates/ExerciseUpdateHandler.java [hunk 5]
```diff
@@ -22,23 +15,25 @@
...
-    public ExerciseUpdateHandler(File cacheFile, TmcJsonParser jsonParser) throws TmcCoreException {
-        super(jsonParser);
+    public ExerciseUpdateHandler(ExerciseChecksumCache cache, TmcApi tmcApi)
+            throws TmcCoreException {
+        super(tmcApi);
         exerciseChecksums = new HashMap<>();
-        if (cacheFile == null) {
+        if (cache == null) {
             String errorMessage = "ExerciseUpdateHandler requires non-null cacheFile to function";
             throw new TmcCoreException(errorMessage);
```
Comment (jamo) [inline src/main/java/fi/helsinki/cs/tmc/core/communication/updates/ExerciseUpdateHandler.java]: inline error message?

File: src/main/java/fi/helsinki/cs/tmc/core/spyware/DiffSender.java [hunk 6]
```diff
@@ -55,7 +55,6 @@ public HttpResult sendToUrl(byte[] diffs, String url) {
             return urlCommunicator.makePostWithByteArray(
                             url, diffs, createHeaders(), new HashMap<String, String>());
         } catch (IOException ex) {
-            System.err.println(ex.getMessage());
             return new HttpResult(ex.getMessage(), 500, false);
```
Comment (jamo) [inline src/main/java/fi/helsinki/cs/tmc/core/spyware/DiffSender.java:58]: We maybe should also log this -- and make sure we treat the snapshot data correctly

File: src/test/java/fi/helsinki/cs/tmc/core/commands/RunCheckStyleTest.java [hunk 7]
```diff
@@ -33,48 +28,33 @@
...
 
     @Test
-    public void testRunCheckStyle() throws Exception {
-        Path path = Paths.get("asdf", "asdf", "qwerty", "kj283u-dkjda93");
-        command.runCheckStyle(path);
-        verify(executorMock, times(1)).runCheckCodeStyle(eq(path));
+    public void testCommandDelegatesToTmcLangs() throws Exception {
+        new RunCheckStyle("somePath", tmcLangsMock).call();
+
+        verify(tmcLangsMock, times(1)).runCheckCodeStyle(eq(Paths.get("somePath")));
```
Comment (jamo) [inline src/test/java/fi/helsinki/cs/tmc/core/commands/RunCheckStyleTest.java]: the times(1) is probably not needed.

File: src/test/java/fi/helsinki/cs/tmc/core/communication/TmcApiTest.java [hunk 8]
```diff
@@ -104,7 +105,7 @@ public void parsesPasteUrlFromJson() throws IOException, TmcCoreException {
                 .thenReturn(fakeResult);
         assertEquals(
                 "https://tmc.mooc.fi/staging/paste/ynpw7_mZZGk3a9PPrMWOOQ",
-                tmcJsonParser.getPasteUrl(fakeResult));
+                tmcApi.getPasteUrl(fakeResult));
     }
 
     String realAddress = "http://real.address.fi";
```
Comment (jamo) [inline src/test/java/fi/helsinki/cs/tmc/core/communication/TmcApiTest.java:111]: Let's prefer exmple.com as it's sole purpose is usage like this.

File: src/test/resources/json/courses.json [hunk 9]
```diff
@@ -4,10 +4,10 @@
...
             "id": 3,
             "name": "2013_ohpeJaOhja",
-            "details_url": "https://tmc.mooc.fi/staging/courses/3.json",
-            "unlock_url": "https://tmc.mooc.fi/staging/courses/3/unlock.json",
-            "reviews_url": "https://tmc.mooc.fi/staging/courses/3/reviews.json",
-            "comet_url": "https://tmc.mooc.fi:8443/comet",
+            "details_url": "http://localhost:8080/courses/3.json",
+            "unlock_url": "http://localhost:8080/courses/3/unlock.json",
+            "reviews_url": "http://localhost:8080/courses/3/reviews.json",
+            "comet_url": "http://localhost:8443/comet",
```
Comment (jamo) [inline src/test/resources/json/courses.json:10]: or example.com

**Comment by jamo** [comment]
Nice work, LGTM (see comments above)

File: src/main/java/fi/helsinki/cs/tmc/core/spyware/DiffSender.java (hunk 6 above)
Comment (ljleppan) [inline src/main/java/fi/helsinki/cs/tmc/core/spyware/DiffSender.java:58]: Logging will be added at a later date

File: src/test/resources/json/courses.json (hunk 9 above)
Comment (ljleppan) [inline src/test/resources/json/courses.json]: These are pointing to wiremock, example.com won't do.

**Pull request merged**